### ISSUE TITLE
python: libdnf5.repo.RepoConfig.baseurl values are not expanded

### ISSUE THREAD

**Issue by gotmax23** [rpm-software-management/dnf5#306]
Take the following code:

``` python
#!/usr/bin/env python3

import libdnf5

BASEURL = (
    "https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/$basearch/"
)
BASEURLE = "https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/x86_64/"


def main():
    for index, baseurl in enumerate(([BASEURL], [BASEURLE])):
        base = libdnf5.base.Base()
        base.load_config_from_file()
        base.setup()
        repo = base.get_repo_sack().create_repo(f"repro-test{index}")
        repo.get_config().baseurl().set(baseurl)
        base.get_repo_sack().update_and_load_enabled_repos(False)
        package = next(iter(libdnf5.rpm.PackageQuery(base)), None)
        print(f"{baseurl}: {package.get_nevra() if package else package}")


if __name__ == "__main__":
    main()
```

```console
$ ./libdnf5_repos.py
['https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/$basearch/']: None
['https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/x86_64/']: 389-ds-base-2.2.4-3.el9.x86_64
```

As shown, setting the `baseurl` to a value that include `$vars` and needs to be expanded does not work with libdnf5 and causes repos to fail to load. This should be fixed (i.e. vars should automatically be expanded) or the correct way to do this should be documented. This works fine with dnf4:

``` python
#!/usr/bin/env python3

import dnf.base

BASEURL = (
    "https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/$basearch/"
)
BASEURLE = "https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/x86_64/"


def main():
    for index, baseurl in enumerate(([BASEURL], [BASEURLE])):
        base = dnf.base.Base()
        for repo in base.repos.iter_enabled():
            repo.disable()
        base.repos.add_new_repo(f"repro-test{index}", base.conf, baseurl=baseurl)
        base.fill_sack()
        print(f"{baseurl}: {next(iter(base.sack.query()))}")


if __name__ == "__main__":
    main()
```
``` console
$ ./dnf_repos.py
['https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/$basearch/']: AtomicParsley-0.9.5-20.fc37.x86_64
['https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/x86_64/']: AtomicParsley-0.9.5-20.fc37.x86_64

```

**Comment by j-mracek** [comment]
With the request there is an additional requirement - load vars before repo is created - what about to lock more the workflow?

**Comment by gotmax23** [comment]
> what about to lock more the workflow?

I don't parse. Can you rephrase? The current docs already say to run `base.setup()` before adding repositories.

**Comment by gotmax23** [comment]
I've implemented this myself for now: https://git.sr.ht/~gotmax23/fedrq/commit/e84d2379884ef38c999f09f4182b925e22fd4690

**Comment by j-mracek** [comment]
The problem is that configuration setter is a simple method that is not intended to modify provided values.

But the issue could be resolved by multiple ways

- [x]  Provide a workaround
You can use vars substitution when you want to substitute variables in the string
```
#!/usr/bin/env python3

import libdnf5

BASEURL = (
    "https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/$basearch/"
)
BASEURLE = "https://kojihub.stream.centos.org/kojifiles/repos/c9s-build/latest/x86_64/"


def main():
    for index, baseurl in enumerate(([BASEURL], [BASEURLE])):
        base = libdnf5.base.Base()
        base.load_config_from_file()
        base.setup()
        vars = base.get_vars().get()
        base_subst = [vars.substitute(baseurl[0])]
        repo = base.get_repo_sack().create_repo(f"repro-test{index}")
        repo.get_config().baseurl().set(base_subst)
        base.get_repo_sack().update_and_load_enabled_repos(False)
        package = next(iter(libdnf5.rpm.PackageQuery(base)), None)
        print(f"{baseurl}: {package.get_nevra() if package else package}")


if __name__ == "__main__":
    main()
```

- [ ] Document workaround
- [ ] Create factoru method for repository that will substitute all provided strings by default

**Comment by gotmax23** [comment]
> Create factory method for repository that will substitute all provided strings by default

A factory function like dnf has would indeed be useful.
